### PR TITLE
Template loop bounds for flexible evaluate/integrate function

### PR DESCRIPTION
**Pull request by bergbauer**
This reduces the instruction count of 
- `internal::evaluate_tensor_product_value_and_gradient()` 
- `internal::integrate_add_tensor_product_value_and_gradient()`

and therefore makes the `FEPointEvaluation::evaluate()/integrate()` functions faster.

@kronbichler

### CODE REVIEW

**Review by peterrum — APPROVED** [review]
/rebuild

File: include/deal.II/matrix_free/tensor_product_kernels.h
Comment (kronbichler) [inline include/deal.II/matrix_free/tensor_product_kernels.h:3160]:
```suggestion
        // Generate separate code with known loop bounds for the most common cases
        if (n_shapes == 2)
          inner_result = do_interpolate_xy<dim, 2, Number2, Number>(
            values, renumber, shapes, n_shapes, i);
```

File: include/deal.II/matrix_free/tensor_product_kernels.h
Comment (kronbichler) [inline include/deal.II/matrix_free/tensor_product_kernels.h:3301]: I like the fact that you do not use `integrate`, as this is really just the addition of a quantity multiplied by all test functions, accumulated into a field. But I think `test` alone is not expressive enough, so maybe
```suggestion
  do_apply_test_functions_xy(AlignedVector<Number2> &                    values,
```

File: include/deal.II/matrix_free/tensor_product_kernels.h
Comment (kronbichler) [inline include/deal.II/matrix_free/tensor_product_kernels.h:3306]: To get this more compact, and make it more similar to the `do_interpolate_xy`, you could also collect these three values in an `std::array<Number2, 3>`. This would reduce the arguments to pass in, possibly making them fit in a single line.

**Review by kronbichler — COMMENTED** [review]
This mostly looks good to me. I have some stylistic comments. I should not approve this because I was involved in the development.

File: include/deal.II/matrix_free/tensor_product_kernels.h
Comment (drwells) [inline include/deal.II/matrix_free/tensor_product_kernels.h:2986]: Does this need to be `DEAL_II_ALWAYS_INLINE`?

File: include/deal.II/matrix_free/tensor_product_kernels.h
Comment (drwells) [inline include/deal.II/matrix_free/tensor_product_kernels.h:2984]: Could you add some documentation? The meaning of the function name isn't obvious to me.

**Review by drwells — CHANGES_REQUESTED** [review]

File: include/deal.II/matrix_free/tensor_product_kernels.h
Comment (bergbauer) [reply include/deal.II/matrix_free/tensor_product_kernels.h:2984]: I have added a small comment. Do you think this is enough?

File: include/deal.II/matrix_free/tensor_product_kernels.h
Comment (bergbauer) [reply include/deal.II/matrix_free/tensor_product_kernels.h:2989]: What speaks against `DEAL_II_ALWAYS_INLINE` from your perspective? It definitely should be inlined, the goal of this PR is to improve this performance-critical part of the code.

File: include/deal.II/matrix_free/tensor_product_kernels.h
Comment (bergbauer) [reply include/deal.II/matrix_free/tensor_product_kernels.h]: Done and yes it fits! I have added comments on which entry is which.

File: include/deal.II/matrix_free/tensor_product_kernels.h
Comment (peterrum) [inline include/deal.II/matrix_free/tensor_product_kernels.h:3392]: Instead of spelling out all the cases, you could also write:

```diff
diff --git a/include/deal.II/matrix_free/tensor_product_kernels.h b/include/deal.II/matrix_free/tensor_product_kernels.h
index e3a5c35..420b996 100644
--- a/include/deal.II/matrix_free/tensor_product_kernels.h
+++ b/include/deal.II/matrix_free/tensor_product_kernels.h
@@ -3296,7 +3296,7 @@ namespace internal
 
 
 
-  template <int dim, int length, typename Number2, typename Number>
+  template <int dim, typename Number2, typename Number, int length = 2>
   inline DEAL_II_ALWAYS_INLINE void
   do_test_xy(AlignedVector<Number2> &                    values,
              const std::vector<unsigned int> &           renumber,
@@ -3307,7 +3307,22 @@ namespace internal
              const int                                   n_shapes_runtime,
              int &                                       i)
   {
-    const int n_shapes = length > 0 ? length : n_shapes_runtime;
+    constexpr int max_length = 6;
+
+    if (n_shapes_runtime != length)
+      return do_test_xy<dim,
+                        Number2,
+                        Number,
+                        std::min(max_length + 1, length + 1)>(values,
+                                                              renumber,
+                                                              shapes,
+                                                              test_grad_x,
+                                                              test_grad_y,
+                                                              test_value_z,
+                                                              n_shapes_runtime,
+                                                              i);
+
+    const int n_shapes = length > max_length ? length : n_shapes_runtime;
     for (int i1 = 0; i1 < (dim > 1 ? n_shapes : 1); ++i1)
       {
         const Number2 test_value_y =
@@ -3377,60 +3392,14 @@ namespace internal
                                       gradient[1] * shapes[i2][0][2] :
                                       (dim > 1 ? gradient[1] : Number2());
 
-        if (n_shapes == 2)
-          do_test_xy<dim, 2, Number2, Number>(values,
-                                              renumber,
-                                              shapes,
-                                              test_grad_x,
-                                              test_grad_y,
-                                              test_value_z,
-                                              n_shapes,
-                                              i);
-        else if (n_shapes == 3)
-          do_test_xy<dim, 3, Number2, Number>(values,
-                                              renumber,
-                                              shapes,
-                                              test_grad_x,
-                                              test_grad_y,
-                                              test_value_z,
-                                              n_shapes,
-                                              i);
-        else if (n_shapes == 4)
-          do_test_xy<dim, 4, Number2, Number>(values,
-                                              renumber,
-                                              shapes,
-                                              test_grad_x,
-                                              test_grad_y,
-                                              test_value_z,
-                                              n_shapes,
-                                              i);
-        else if (n_shapes == 5)
-          do_test_xy<dim, 5, Number2, Number>(values,
-                                              renumber,
-                                              shapes,
-                                              test_grad_x,
-                                              test_grad_y,
-                                              test_value_z,
-                                              n_shapes,
-                                              i);
-        else if (n_shapes == 6)
-          do_test_xy<dim, 6, Number2, Number>(values,
-                                              renumber,
-                                              shapes,
-                                              test_grad_x,
-                                              test_grad_y,
-                                              test_value_z,
-                                              n_shapes,
-                                              i);
-        else
-          do_test_xy<dim, -1, Number2, Number>(values,
-                                               renumber,
-                                               shapes,
-                                               test_grad_x,
-                                               test_grad_y,
-                                               test_value_z,
-                                               n_shapes,
-                                               i);
+        do_test_xy<dim, Number2, Number>(values,
+                                         renumber,
+                                         shapes,
+                                         test_grad_x,
+                                         test_grad_y,
+                                         test_value_z,
+                                         n_shapes,
+                                         i);
       }
   }
 
```

**Comment by peterrum** [comment]
> This reduces the instruction count of

Have you done something additional to fixing the loop bounds?

File: include/deal.II/matrix_free/tensor_product_kernels.h
Comment (drwells) [reply include/deal.II/matrix_free/tensor_product_kernels.h:2989]: We typically don't force inlining (i.e., we let the compiler decide) unless the function is very small - I would expect using `DEAL_II_ALWAYS_INLINE` here to actually make things slower. It might be worth trying and measuring to see which is faster.

File: include/deal.II/matrix_free/tensor_product_kernels.h
Comment (kronbichler) [reply include/deal.II/matrix_free/tensor_product_kernels.h:3392]: This is based on my suggestion, so let me explain. If you call this recursively, then the compiler generates a function prologue/epilogue for every step that is not inlined. Given that we should assess the suggestion of @drwells at https://github.com/dealii/dealii/pull/14972#discussion_r1148656523, I would like to ensure that performance is not affected. The current code is based on the experience of `FEEvaluationImpl`, where the effect of stepping into another function to arrive at the target degree is quite visible especially in 2D (around 10% of the instructions, which is pretty bad). Note that this is an inner loop component that does less work than `FEEvaluation` in 2D in all cases, so picking a good strategy is vital. It might be worth measuring for some representative cases, but I would rank the ability to avoid `always_inline` higher than repeating these steps, especially if we manage to fit the call in one or two lines. :wink:

File: include/deal.II/matrix_free/tensor_product_kernels.h
Comment (kronbichler) [reply include/deal.II/matrix_free/tensor_product_kernels.h:2989]: I agree, we should take a look. It depends on how much function prologue/epilogue is necessary and how the compiler decides to realize the inlining. Compiler can make pretty bad decisions sometimes, because it is hard to know what is really an inner loop (thus worth inlining) and what gets called less often and should limit the code size instead (better not to inline). The current suggestion is based on past experience with compilers, so it might be different here. @bergbauer can you please collect the run time for your benchmark and a range of degrees with the two options? Either way, we should use something like https://github.com/dealii/dealii/blob/b8135fa6eb6e8b59cf9aab7b5e2ded518d2fdf61/include/deal.II/matrix_free/evaluation_kernels_hanging_nodes.h#L31-L35

Comment (bergbauer) [reply include/deal.II/matrix_free/tensor_product_kernels.h:2989]: Screenshot from 2023-03-27 11-12-56

I collected the throughput and instruction counts for the two relevant functions, you cannot see much of a difference between `inline` and `DEAL_II_ALWAYS_INLINE`. What do you think, should we go for plain `inline`?

Comment (kronbichler) [reply include/deal.II/matrix_free/tensor_product_kernels.h:2989]: The difference seems to be in the noise. So let us go for the code without `DEAL_II_ALWAYS_INLINE` as suggested by @drwells, and only add in case we have a case with a problem. Thank you for measuring the performance.

**Comment by bergbauer** [comment]
>Have you done something additional to fixing the loop bounds?

No, this is the only change in this PR.

>How much are we talking about?

I measure about 5-10% higher throughput in my benchmark depending on the polynomial degree.

@peterrum

File: include/deal.II/matrix_free/tensor_product_kernels.h
Comment (bergbauer) [reply include/deal.II/matrix_free/tensor_product_kernels.h]: Thanks for making me aware of the difference between `inline` and `DEAL_II_ALWAYS_INLINE`. Done.

**Review by kronbichler — COMMENTED** [review]
Looks good to me.

**Comment by bergbauer** [comment]
Do you think this is ready now? @peterrum @drwells I already have a follow-up patch ready, depending on this PR.

**Review by peterrum — APPROVED** [review]

**Review by tjhei — APPROVED** [review]
👍

**Comment by tjhei** [comment]
> I measure about 5-10% higher throughput in my benchmark depending on the polynomial degree.

This is awesome. 👍

**Review by masterleinad — APPROVED** [review]
Looks OK to me.

**Review by drwells — APPROVED** [review]

**Pull request merged**
